### PR TITLE
server: new health report facility

### PR DESCRIPTION
**Pull request by knz**
Supersedes #59191 

tldr: nodes now collects a health report periodically in the
background, and report it as payload for HTTP GET on `/health`.

This work stems from a desire to inspect the state of subsystems
inside a crdb node, so as to assess its general health. The binary
result previously provided by `/health` was deemed insufficient.

To achieve this, this patch extends the "node alerts" prototype
feature from 2018. This feature used thresholds on certain node-level
metrics and generates "gossip alerts" (visible to other nodes, via
`crdb_internal.gossip_alerts`) when thresholds are exceeded.

The extension is threefold:

- the results of the node alerting code is now preserved in-RAM
  outside of gossip, so it can be reported by `/health`.
- the values for all the "tracked metrics" that are relevant for
  alerts are now included in the health report, not just
  the alerts; they are reported even when their value is not in
  a danger zone (i.e the health report contains data even if
  there is no problem currently detected).
- the abstraction is generalized from a system originally
  specialized to *node* metrics, to one that generally can assess
  the health of a server.

To the benefit of this last point, the code that implemented the
previous feature was extracted from the `writeNodeStatus` logic,
so that it become less intimately linked to the ability to persist
the `NodeStatus` proto in KV.

This way, the health checker can now maintain, in-RAM, a periodically
updated health report, even when the node status range is not
available to write in KV.

Release note (api change): The `/health` API (public) is now separate
from `/admin/_v1/health` (internal to the DB console). The `/health`
endpoint also now returns a report of the latest measured health of
the node.

### CODE REVIEW

**Comment by knz** [comment]
The plan before this merges is to extend the mechanism with SQL-level checks:

- availability of system tables
- availability of authentication
- availability of table leases
- (maybe) job progress
- more depending on interest and complexity

**Comment by tbg** [comment]
How do you think the SREs will/should consume this? Would they initiate pages based on certain health responses or on any alerts? Or would they inspect the health endpoint early when responding to a page just to get their bearings on possible problems in the cluster?
And, since we are adding this to the public API, isn't it immediately going to calcify, which may also not be useful since our primary audience are our SREs?

**Comment by knz** [comment]
> How do you think the SREs will/should consume this?

Right now my first next goal is to have this data available for troubleshooting problems.

(Which reminds me, I probably want to store a copy of the entire health report in the node status. Will do this in a later commit.)

> Would they initiate pages based on certain health responses or on any alerts?

Yes, I think that will be a good idea.

> Or would they inspect the health endpoint early when responding to a page just to get their bearings on possible problems in the cluster?

That, too.

> And, since we are adding this to the public API, isn't it immediately going to calcify, which may also not be useful since our primary audience are our SREs?

I don't understand your concern? What's calcifying?

**Comment by tbg** [comment]
> I don't understand your concern? What's calcifying?

Can we easily change and remove it once it's in a release? I don't think we are ready to fully commit to this approach, though I'd like to try it out. As mentioned, I primarily want to make life easier for the SREs. If they can't work with this approach, I think the approach needs to change.

**Comment by knz** [comment]
> Can we easily change and remove it once it's in a release?

IMHO yes:

1. `/health` is not a versioned API and thus does not come with guarantees about the specifics of its response (only whether it suceeds or not)
2. if we do create a versioned API e.g. `/api/v2/healthreport`, then we can change it by adding a new version under `/api/v3`. That was the entire point of the API versioning project.

**Comment by joshimhoff** [comment]
Moving email discussion to the PR!

> tbg: I think more concretely the question is how we can build "alertability" into CockroachDB in a way that the SREs would like to consume

What we want to consume is time-series metrics which we can use prometheus to store over time & aggregate, alertmanager to alert on, and grafana to visualize. All these aspects are critical; we don't alert on a single failure; we alert on a high error rate over multiple nodes over ~1 hour; the way the math works out this leads to a high signal to noise alert that still fires quickly enough in case of global outage given our pager SLA (see https://github.com/cockroachlabs/managed-service/pull/2991#discussion_r435287100). Prometheus makes this kind of thing straight-forward, & the SREs know the black magic of tuning alerting configs. Relatedly, I think the alerting thresholds needs to be changeable by SREs without requiring a CRDB release; we need a lot of control over those thresholds to manage our pager volume; the actual alert definitions need to be SRE-owned code / config. One last benefit I can think of is that prometheus & CRDB need not share a failure domain (they can run on different k8s clusters for example), so monitoring can work well even during outages; CRDB just needs to be scrap-able (pretty low bar).

Basically, there is an existing tool for this job widely used in the industry for a bunch of good reasons IMHO & we should use it. It provides a nice boundary where SWE / SRE expose metrics & then SRE owns alerting on those metrics (with consultation from SWE certainly).

> tbg: what corners we can cut to give you something in the short term

No need to cut corners here IMHO. We should steadily expand the number of useful time-series metrics that measure aspects of the customer experience over time. Probers, white-box metrics, etc. If DB eng thinks SRE should alert on it (alerting != monitoring), then that's awesome, we love alerting, and we badly need the help of DB eng in understanding the internals of the system well enough to have good alerting coverage, But also plz talk to SRE, as we are very careful about keeping our alerting setup from being too noisy!

If SRE seems picky about alerting, then yes we are. We are bc/ our skin is very much in the game. It's a black art keeping the pager load low.

Also though I'm arguing agains this approach to alerting, that doesn't mean I'm against the specific checks being added.

**Comment by tbg** [comment]
Thanks for the info! I don't mind the being "picky about alerting" at all. I think mostly what we have been looking for is a right way of doing things - looks like the SREs have strong opinions on that - and that's great. Is there something we can learn from in terms of how to actually implement this? i.e. what do the timeseries that alerting is based on look like? I assume either event counters or gauges are fine? It almost seems like the real disconnect is in explaining what metrics are important for alerting and what reasonable thresholds are, and adding new metrics where we see blind spots.

There's still a question on how this can carry over to non-CC deployments. Everyone deploying CRDB is struggling with alerting, and ideally we solve at least a basic part of that problem for all. Maybe we just need a process for translating the SRE alerting rules into something we make part of our public deployment story at some interval? I assume we can't just throw them over the wall and expect things to work.

**Comment by knz** [comment]
The way I've been thinking about this is as follows:

1. continue developing the health report in this PR, because we want this data in `debug zip` anyway
2. this code is able to generate alert events on certain "edge" or "level" situations, eg a metric exceeding a threshold or a system not being available. Today these events are propagated to the rest of the cluster via gossip, and IMHO this is a good thing because this way we can pick the alert up even if the problem node goes down on its own.
3. **what is not done yet which I intend to add**: push a structured event with these alerts through logging. We now (or soon will) monitor health events through our networked logging infra. If we had health alerts in there, we could have the log monitoring equipped with paging on certain logging events (or groups thereof across multiple nodes).
4. then we can teach on-prem customers how to do the same.

How does that look like?

**Comment by tbg** [comment]
I'm still worrying that we're walking down a muddied road. I think Josh is telling us: SREs don't care about this approach for alerting, just give us metrics we can subscribe to, and in particular don't hard-code the thresholds in CRDB. He's fine with more information added to the health check responses, but doesn't particularly care about them. So we're not doing any of this for the SREs, as far as I can tell, which I consider problematic. I would rather do something that benefits the SREs too, but I think they just have better tools available than we could provide internally, and will prefer to use them. This work is aimed at improving our customer success' interactions with customers by providing alerting to *them* when being pulled into customer issues. Is that a fair characterization of the goals? By maintaining hard-coded alerts internally, we do benefit SREs to some degree, as they can check those against their alerting periodically, so it's not entirely disconnected.

The one point I would make is that it seems like it would be a mistake if the health checker actually started any work to check health. From the SRE point of view, I am deriving that subsystems that check health should do this at their leisure and report directly to metrics. The health checker then looks *only* at metrics. There is a question about observability. If, say, some range probe fails, wouldn't you want to know details about the probe? Instead, you get an incremented metric counter. Is the solution to go to the logs? Does each type of probing need a ring buffer that can be queried for details? There seems to be a little bit of infrastructure and direction we should decide on.

I also think the way the metrics are pulled from the node status is a little ad-hoc (I wrote it back then) but it seems to have worked ok. I don't like that it uses random strings to link these metrics together; perhaps that's fine though as long as we have a test but I don't think it particularly incentivizes alerting as a first-class citizen. You could imagine that the alertability of a metric could be part of the metrics metadata instead, and thus become part of the catalog (which in theory would allow us to auto-generate prometheus alerting rules to start users off), so that there wouldn't be this random map in the codebase that you have to update. Down the road, in theory, we could maintain a system table that contains a mapping from metric to alerting threshold, which gives a limited, internal version of the more powerful tooling SREs use via prometheus.

I'm not saying this all has to happen, but I'd like a clear opinion the direction.

**Comment by joshimhoff** [comment]
> continue developing the health report in this PR, because we want this data in debug zip anyway

> This work is aimed at improving our customer success' interactions with customers by providing alerting to them when being pulled into customer issues. Is that a fair characterization of the goals?

I think the above two points are big; they significantly clarify things for me! Yes: SRE wants to alert on data in prometheus (or similar system). And CRL never receives **just** a debug zip from a CC customer. OTOH, I can see why you'd want info-only alerts in that debug zip, as it automates work that you have to do manually today.

In the cloud product, we have a richer set of tools at our disposal, and this leads to a certain amount of conflict re: direction.

I think a key thing is: There is a difference between an info-only alert (obserability) & a page (or ticket). The latter is what I mean by alerting. Those needs to be in SRE's control. We select / tune the alerting conditions carefully and we need to know rather exactly what the alerts mean. Info-only alerts OTOH are a useful obserability improvement, and the stakes re: well-tuned alert that SRE understands are way way way lower. An example that exists today is problem ranges according to the admin UI. We don't get woken up at 3am for >0 problem ranges, but they are useful obserability tool definitely, when another alert does wake you up.

> There's still a question on how this can carry over to non-CC deployments. Everyone deploying CRDB is struggling with alerting, and ideally we solve at least a basic part of that problem for all. Maybe we just need a process for translating the SRE alerting rules into something we make part of our public deployment story at some interval?

Ya I think that's wise. Another thing that will help eventually is the public k8s operator, tho you gotta be on k8s!

> i.e. what do the timeseries that alerting is based on look like? I assume either event counters or gauges are fine? It almost seems like the real disconnect is in explaining what metrics are important for alerting and what reasonable thresholds are, and adding new metrics where we see blind spots.

These are good Qs. Yes re: counters & gauges. I think easier topic to take up in a 1:1 setting at least to start. I think we can do a better job explaining how we think about alerting and engaging more people in the topic, and I expect we will keep working on this moving forward! I want to repeat again that obserability is a place where we are significantly less picky; the more obserability the better!

> If we had health alerts in there, we could have the log monitoring equipped with paging on certain logging events (or groups thereof across multiple nodes).

@knz, could we push plain events to the log? No threshold applied leading to an alert in the log, just the event. The log can be consumed to create alerts. SRE can do this as they'd like to via splunk cloud. Debug zip can do it automatically in order to surface info-only alerts to CRL eng working with an on-prem customer. If the threshold is applied before hitting the splunk log, I don't think SRE will alert (page) on it.

**Comment by knz** [comment]
> could we push plain events to the log?

yes of course

**Comment by knz** [comment]
Ok I am glad we are chatting about this. From my perspective, the convo above can be synthetized in three "high level" concepts:

1. which entity/program/system is responsible for translating metrics/conditions into "notable events" (i.e. that's something interesting is happening), previously called "alerts" inside crdb
2. which entity/program/system is responsible for translating metrics/conditions into alerts/pages (i.e. an invitation for a SRE to actually do something)
3. which data should be capture so that investigations become easier.

What you Josh have now repeated multiple times, if my understanding is right:

- you don't care about points (1) and (3) as much as you care about point (2). I would still like to hear your thoughts about points (1) and (3) but we can discuss separately.
- regarding point (2) (and also point (1) to some extent), you strongly recommend that the responsibility for translating data into alerts/pages should not be taken by crdb itself. I am in very strong agreement about this.

Then Tobias says that if we build something for point (2) outside of crdb, we should take care of ensuring that logic is exportable so our on-prem users can have it too. Yes, I'd agree with that too.


Now, in the interest of v21.1 (and a backport to v20.2) I would still like to move forward with something that helps, if only just a little, for all 3 points above. I have this in mind:

1. have a runtime-configurable list of metrics that the health checker would monitor (in addition to the prometheus endpoint), instead of a static list
2. create additional conditions/"metrics" inside the code for things that are neither counters nor gauges, or perhaps would evolve dynamically over time (our *metric* set is fixed at build time):
   - availability of system tables / ranges
   - availability of certain sub-systems inside SQL and KV
   - availability of network services
3. make the health checker inside crdb report changes to all these values into structured logging events, without judgement as to whether they are alert-worthy or not; so that we can build alerting/paging from our network logging infrastructure.

(That would come in addition to the current logic in this PR which also aims to produce a health *report* about all these metrics/conditions in a place we can look at when troubleshooting, with the goal to aid with point (3) above.)

Thoughts?

**Comment by joshimhoff** [comment]
Thank you for the excellent summary. I am happy we are talking about this too! I don't feel I concretely understand the 21.1 proposal yet so asking some more Qs. I have a few more thoughts also.

> have a runtime-configurable list of metrics that the health checker would monitor (in addition to the prometheus endpoint), instead of a static list

Here we are talking about an HTTP health check? The HTTP health check would check certain metrics? I might be missing some context re: what the health checker is at a high level.

> create additional conditions/"metrics" inside the code for things that are neither counters nor gauges, or perhaps > would evolve dynamically over time (our metric set is fixed at build time):
> - availability of system tables / ranges
> - availability of certain sub-systems inside SQL and KV
> - availability of network services

While I can certainly imagine a condition / "metric" that can't be represented by a counter or gauge or even histogram, I don't see why availability of a system table / range or any of the other mentioned items is one such thing. Seems to fit quite naturally into the prometheus data model. I'd expect a request & an error counter, which can be divided to compute availability.

> make the health checker inside crdb report changes to all these values into structured logging events, without judgement as to whether they are alert-worthy or not; so that we can build alerting/paging from our network logging infrastructure.

I still have a pretty strong preference to (mostly) lean on a proper time-series monitoring system like prometheus for alerting and to a certain extent monitoring. I don't object to events being pushed to network logging infra in the least bit but I do worry if the intention in doing this is to focus on events pushed to network logging *over* time-series metrics. I am not sure that's being proposed though really. I get the sense the desire to use network logging has more to do with the "make ops on on-prem easy" goal than a principled general preference in favor of events pushed to network logging over time-series metrics?

I also am still reflecting on *why* I prefer time-series metrics to events pushed to network logging infra. I have some ideas about that but I am going to think a bit more & maybe talk to some teammates too.

Lastly, I think re: 21.1 we should bias for action while keeping this conversation going so we can agree on a good general direction that benefits cloud & on-prem!

**Comment by daniel-crlabs** [comment]
>a casual inspection of the customer's request suggests they should use /health?ready=1 as probe

This is what they were doing and it was causing problems, this is why they created their own custom readiness check:

>Once the readiness passes - the node will become available in our load balancer and will start accepting SQL connections. However, almost all queries sent to that node will either time out(we have 5s application timeout on queries) or take unusually long time to complete. This behavior is quite disruptive to our product and customers.
>
>To mitigate that problem we have a custom readiness check in place that would check the node's metrics ( range_snapshots_normal_applied, range_snapshots_learner_applied ) and keep the node in "not ready" state while snapshots are applying.

Their question:

>We are looking to reduce the maintenance cost on our side and remove as much of our in-house config as possible. 

They do not want to have to use a custom health check as they have to do at the moment.

>Can you advise the best way to manage node readiness so that we don't see increased latencies while it's syncing? Is it something that can be solved in the official helm charts or in CRDB itself?

Do we have an update/suggestion I can provide the customer? Do I tell them to keep doing what they are doing, as we won't be able to work on this?

**Comment by daniel-crlabs** [comment]
@knz thank you for your feedback, the ticket you are asking about was linked a few comments ago (https://github.com/cockroachdb/cockroach/pull/59656#issuecomment-947007851). With that being said, I'll go ahead and create a separate github issue for this. Thank you for your input.
